### DISCUSSION TITLE
Cross-platform porting

### DISCUSSION THREAD

**Discussion by dsharlet** [dsharlet/LiveSPICE#222] [General]
There are two issues open about porting to linux or mac:
- https://github.com/dsharlet/LiveSPICE/issues/51
- https://github.com/dsharlet/LiveSPICE/issues/202

I'm going to close these and any future issues about porting to other platforms as duplicates and point them here for further discussion.

**Comment by dsharlet** [comment]
I didn't close this issue as a duplicate, but this is also quite relevant: https://github.com/dsharlet/LiveSPICE/issues/203

**Comment by evan-olcott** [comment]
I'm just going to spitball in order to start the conversation, but we may be able to identify different parts of the app and address them differently:

The Math: The Computer Algebra library looks very interesting, but can it be translated to lower-level libraries like BLAS or platform-specific wrappers (like Accelerate on macOS) to get some more power out of it? May be obvious, but I'm only asking because I _really_ don't know.

The Analysis/Compilation: Would it be fair to wait on an optimization phase for something like this before addressing how this would work in a more generic way? I'm not sure I truly understand Linq, but if really we're just coming up with a bunch of coefficients after an analysis phase -- maybe it's easier than I think?

The Audio: Audio processing libraries are different for every platform, but most use the pull-IO style at their root. Each platform will have to have its own code to wrap this, but it's almost boilerplate these days.

The UI: Again, platform-specific to be honest about it. Sure there's some libraries for taking React or Java and making something cross-platform, but the UX is almost always sub-par as a result. To really get engaged properly with it, they need to be natively written, but certainly should have shared UX (how to build a circuit, what parts are available, etc.). Document format plays into this as well - should be something all platforms can read from equally.

For the sake of transparency, my specialties are the latter two in this list, but I'm brushing up on my Linear Algebra in order to be more sensitive to the former and maybe even contribute to it.

**Reply by dsharlet** [reply]
> The Math: The Computer Algebra library looks very interesting, but can it be translated to lower-level libraries like BLAS or platform-specific wrappers (like Accelerate on macOS) to get some more power out of it? May be obvious, but I'm only asking because I really don't know.

It's a pretty different kind of math library. A more comparable alternative would be SymPy.

The main idea behind LiveSPICE that enables huge speedups (orders of magnitude) is to generate a symbolic description of the circuit simulation, and then partially solve that at "compile" time. 

There is a small part of LiveSPICE simulations (during runtime) that could be BLAS calls, but they're small and not that interesting, only a few lines of code. 

https://github.com/symforce-org/symforce is a library that applies a similar idea (and uses SymPy) to a different kind of problem. However, controller math is really very similar to circuit simulation math, so perhaps using that to replace some of LiveSPICE's analysis + symbolic math would work.

I think if I were writing LiveSPICE from scratch today, I'd be looking at SymForce as a dependency for the solver. It doesn't have everything needed here though:
- I can't see any examples of SymForce that do significant symbolic algebraic optimizations between setting up the problem and generating the code for it. It looks to me like the only symbolic computation is generating Jacobians/derivatives. It might still be possible to do this without any SymForce modifications (just grab the Jacobian it generates and generate our partially symbolic solution for Newton's method from that).
- I don't see a JIT compilation backend for SymForce, only static C++ code generators.

> The Analysis/Compilation: Would it be fair to wait on an optimization phase for something like this before addressing how this would work in a more generic way? I'm not sure I truly understand Linq, but if really we're just coming up with a bunch of coefficients after an analysis phase -- maybe it's easier than I think?

I think figuring it out early is important. It is the thing about LiveSPICE that is unique and makes it work. When I worked on LiveSPICE the first time around, this was the core of the project that I spent a lot of time figuring out how to do.

I wrote a description of what is going on here: https://dsharlet.com/2014/03/28/how-livespice-works-numerically-solving-circuit-equations.html. This isn't the only math used in LiveSPICE, but it is the (possibly) novel thing that I had to figure out for myself (at least at the time, this was 10+ years ago).

LiveSPICE's use of LINQ is really quite an abuse of LINQ, which is maybe why it's confusing. LINQ is a way of generating SQL-like queries in .NET languages. It has a backend that can translate the query into low level code and compile it at runtime: https://learn.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression?view=net-8.0 LiveSPICE skips the LINQ front-end and generates these expressions directly itself. This is how LiveSPICE JIT compiles the result of the symbolic simulation solver and turns it into a numerical simulation that is fast.

BTW, you might be wondering... this was needed 10 years, ago, do we still need it now? Circuit simulation is a problem that doesn't parallelize well, so it is mostly a function of single thread performance, which hasn't changed as much as computers in general have. Even typical single thread performance includes SIMD, which we struggle to use. Symbolic algebra generates sparse math, which doesn't work well with SIMD. So LiveSPICE is mostly stuck just running plain old single threaded scalar code.

**Reply by Federerer** [reply]
The 'headless' version (check `Tests` project) should work on Linux/mac as is or with minimal changes.

> The Math: The Computer Algebra library looks very interesting, but can it be translated to lower-level libraries like BLAS or platform-specific wrappers (like Accelerate on macOS) to get some more power out of it? May be obvious, but I'm only asking because I _really_ don't know.

this part is 100% cross-platform and also pretty hard to replace, as we don't have any abstraction layer here.

> The Analysis/Compilation: Would it be fair to wait on an optimization phase for something like this before addressing how this would work in a more generic way? I'm not sure I truly understand Linq, but if really we're just coming up with a bunch of coefficients after an analysis phase -- maybe it's easier than I think?

Computer Algebra produces a `Solution` which contains one or multiple systems of equations, so we can treat this as our generic input layer, or try to do this on lower level by translating produced expression tree to something different (I did this once, generating c code, but it's a little bit limited, as the LINQ expression trees are C#-specific).

> The UI: Again, platform-specific to be honest about it. Sure there's some libraries for taking React or Java and making something cross-platform, but the UX is almost always sub-par as a result. To really get engaged properly with it, they need to be natively written, but certainly should have shared UX (how to build a circuit, what parts are available, etc.). Document format plays into this as well - should be something all platforms can read from equally.

There is a cross-platform flavor of WPF - [Avalonia](https://avaloniaui.net/), which can be used. It is doable, but also labor intensive ;)

**Comment by Federerer** [comment]
> Circuit simulation is a problem that doesn't parallelize well, so it is mostly a function of single thread performance, which hasn't changed as much as computers in general have. Even typical single thread performance includes SIMD, which we struggle to use. Symbolic algebra generates sparse math, which doesn't work well with SIMD. So LiveSPICE is mostly stuck just running plain old single threaded scalar code.

A very good summary 😁

**Comment by evan-olcott** [comment]
Didn't see the headless version in the `Tests`! Very interesting, those will come in very handy!

OK, so it sounds like the ComputerAlgebra aand other math portions are kind of an as-is deal, but I'm still unclear what the "package" the compiler produces is, that becomes the "DSP processor" for the audio. Does it simply take the input in small groups of samples and then... creates a single value from it, uses it as the input value to a series of equations, and comes out with a group of samples for output? What exactly happens at this DSP stage? 

As far as single-threaded performance goes, I totally understand that that's the best way to do these calculations - serially. The audio thread likes doing this as well, and generally runs in its own background thread with a handful of DSP space to do lots of stuff in - there are modern methods that can offload processing onto other threads with minimal latency... having a live passthrough is certainly very important, but even with a latency of 256 samples @ 48k, that's a heck of a lot of time to do some great math on another serial queue.

**Reply by dsharlet** [reply]
> I'm still unclear what the "package" the compiler produces is, that becomes the "DSP processor" for the audio. Does it simply take the input in small groups of samples and then... creates a single value from it, uses it as the input value to a series of equations, and comes out with a group of samples for output? What exactly happens at this DSP stage?

This might be what you are looking for: https://github.com/dsharlet/LiveSPICE/blob/5dda7b729660f1764166472020c8cc656bff9e4d/Circuit/Simulation/Simulation.cs#L156

`_process` is the function that is generated at runtime to implement the simulation. It consumes and produces arrays of samples. `N` is the number of samples we expect in each buffer (256 in your example). Internally, it produces the output samples one at a time as a function of the corresponding input samples. Of course, this function does have internal state that persists across calls, so it is not a pure function.

> ... but even with a latency of 256 samples @ 48k, that's a heck of a lot of time to do some great math on another serial queue.

Definitely... you might be interested in the discussion here: https://github.com/dsharlet/LiveSPICE/discussions/197

**Reply by Federerer** [reply]
> Does it simply take the input in small groups of samples and then... creates a single value from it, uses it as the input value to a series of equations, and comes out with a group of samples for output? What exactly happens at this DSP stage?

Not really, for every input sample it iteratively solves a non-linear system of equations. Essentially the whole circuit boils down to a list of nodes and voltages at those nodes are your unknowns. This uses the previous system state as the initial guess, so you can't really calculate those samples in parallel, because current system state relies on the previous one.

**Comment by evan-olcott** [comment]
OK, digging in to ComputerAlgebra closer -- what was the decision behind using BigIntegers to back the Real structures? Was the precision at the extremities of 64-bit FP values not enough for sound/calculation quality and weighed heavier than the performance costs?

**Reply by dsharlet** [reply]
For a lot of computer algebra purposes, we really need **exact** arithmetic, hence the use of rational numbers with BigInteger as the numerator/denominator. The reason being we are doing symbolic row reduction/gaussian elimination on matrices, and we need those symbolic expressions to cancel exactly. Because the expressions are symbolic, you can't easily determine if the expressions are "small" and can be rounded to 0.

Note that these quantities only exist in the symbolic analysis part of LiveSPICE. Once we are actually running a simulation, these are all converted to doubles, so any slowness from BigInteger only affects solving the circuit, not the simulations themselves.

**Comment by evan-olcott** [comment]
OK, been digging even deeper, finally am able to step-debug through the Tests to see where the calls go...

1. What is a "Visit" and "Visitor"? I'm _guessing_ it's some sort of object that looks at the terms of an expression and converts/formats/reads them in some form for some kind of process, like compiling or displaying...?
2. I think I'm seeing the compiling actions take place, I'm not sure I understand them yet, but it looks like it's kind of .NET or C# based in what the expressions compile to? Is that right?
3. I see a little caching system in there -- I guess I understand it, but how useful does it end up being? Am I seeing a custom Visitor handling the caching?
4. I'm trying to understand the tree of inheritance here, and can you confirm that the "Expression" is basically the fundamental object that, though it in itself doesn't have the values, it hierarchically holds them in various capacities, and it able to be the base point for evaluation?

Thanks in advance

**Reply by dsharlet** [reply]
1. This is an implementation of the visitor pattern: https://en.wikipedia.org/wiki/Visitor_pattern#C#_example. Funnily enough, their example is literally `Expression` and `ExpressionVisitor` :) . We have many visitor implementations for all kinds of processes: compilation and converting to a string as you mentioned, but also many of the core computer algebra operations are implemented as visitors (e.g. differentiation).
2. The expression compiler targets this .NET framework class hierarchy: https://learn.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression?view=net-8.0.
3. In the case of the compiler, the caching is basically a form of CSE (common subexpression elimination). When compiling an expression like `(x + 2) * (x + 2)`, the compiler might generate something like:
```
_1 = x;
_2 = 2;
_3 = x + 2;
_4 = x;
_5 = 2;
_6 = x + 2;
_7 = _3 * _6;
```
But with caching, it would generate something more like this:
```
_1 = x;
_2 = 2;
_3 = x + 2;
_4 = _3 * _3;
```
4. Yes, `Expression` is the root of the computer algebra expression object hierarchy. Things like variables, constants, and arithmetic operations are all part of this class hierarchy.

**Comment by evan-olcott** [comment]
Making some really great progress in a Swift port of this library - I finally have a question I cannot Google around to find an answer:

`public static int LexicalCompareTo<T>(this IEnumerable<T> L, IEnumerable<T> R) where T : IComparable<T>`

How is this C# function helpful, given that IEnumerable does not guarantee order? Wouldn't this give random results basically every time?

**Reply by dsharlet** [reply]
I think it's not possible to enforce an ordering constraint via the type system. There is no generic `IOrderedEnumerable<T>` (there is, but it's not what we want here), and `IList`, would be overly constrained. It probably isn't possible for the framework to do this without a combinatoric explosion of overloads of `Select`, etc.

So basically, while `IEnumerable<T>` does not guarantee order (because it might be a hash table or whatever), it *can* produce a reliable order if it's an ordered container, which I'm assuming it is in the places we use it.

> Making some really great progress in a Swift port of this library - I finally have a question I cannot Google around to find an answer:

Sounds like a fun project, but TBH, I'm personally not excited about a port from one questionably portable language to another questionably portable language. Is anyone using Swift on Linux or Windows? I see it does claim to support those platforms... but so does C#/.NET!

**Comment by Federerer** [comment]
I acquired a new hardware, so there might be some porting effort starting soon 😁 
Initial testing after some quick and dirty edits to remove windows forms dependencies seems promising:
```
Opening Examples/Marshall JCM800 2203 Preamp.schx
Schematic loaded from 'Examples/Marshall JCM800 2203 Preamp.schx'
Building circuit...
Build: 0 errors, 0 warnings
Analyze time: 00:00:00.0065353
Building solution for h=2,2675736961451248E-05
Performing steady state analysis...
Performing transient analysis...
System solved, 2 solution sets for 40 unknowns.
Solve time: 00:00:00.2076877
Newton iteration size: 11
ToReplace: 682, replaced: 682
Build time: 00:00:00.0355155
700 kHz, 15,9x real time
```
